fem/particleset.hpp

```diff
@@ -0,0 +1,521 @@
+// Copyright (c) 2010-2025, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_PARTICLESET
+#define MFEM_PARTICLESET
+
+#include "../config/config.hpp"
+#include "../linalg/linalg.hpp"
+#include "gslib.hpp"
+#include "kernel_dispatch.hpp"
+
+#if defined(MFEM_USE_MPI) && defined(MFEM_USE_GSLIB)
+namespace gslib
+{
+extern "C"
+{
+   struct comm;
+   struct crystal;
+} //extern C
+} // gslib
+#endif // MFEM_USE_MPI && MFEM_USE_GSLIB
+
+namespace mfem
+{
+
+/// Container for data associated with a single particle. See \ref ParticleSet
+/// for more information.
+class Particle
+{
+protected:
+   Vector coords;
+   std::vector<Vector> fields;
+   std::vector<Array<int>> tags;
+public:
+   /** @brief Construct a Particle instance.
+    *  @param[in] dim          Spatial dimension (size of @a coords).
+    *  @param[in] field_vdims  Vector dimensions of particle fields.
+    *  @param[in] num_tags     Number of integer tags.
+    */
+   Particle(int dim, const Array<int> &field_vdims, int num_tags);
+
+   /// Get the spatial dimension of this particle.
+   int GetDim() const { return coords.Size(); }
+
+   /// Get the number of fields associated with this particle.
+   int GetNFields() const { return fields.size(); }
+
+   /// Get the vector dimension of field \p f .
+   int FieldVDim(int f) const { return fields[f].Size(); }
```

Comment:
Maybe `GetFieldVDim()` to be more consistent with the naming of other things here?